src/client_builder.rs

```diff
@@ -61,38 +134,131 @@ where
         ClientBuilder {
             domain,
             port,
-            starttls: false,
+            mode: ConnectionMode::Tcp,
+            #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+            tls_kind: TlsKind::Any,
+            #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+            no_tls_verify: false,
         }
     }
 
     /// Use [`STARTTLS`](https://tools.ietf.org/html/rfc2595) for this connection.
-    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
     pub fn starttls(&mut self) -> &mut Self {
-        self.starttls = true;
+        self.mode = ConnectionMode::StartTls;
+        self
+    }
+
+    /// Use TLS for this connection.
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    pub fn tls(&mut self) -> &mut Self {
+        self.mode = ConnectionMode::Tls;
+        self
+    }
+
+    /// Sets the TLS backend to use for this connection.
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    pub fn tls_kind(&mut self, kind: TlsKind) -> &mut Self {
+        self.tls_kind = kind;
+        self
+    }
+
+    /// Controls the use of certificate validation.
+    ///
+    /// Defaults to `false`.
+    ///
+    /// # Warning
+    ///
+    /// You should only use this as a last resort
+    ///
+    /// See [`native_tls::TlsConnectorBuilder::danger_accept_invalid_certs`],
+    /// [`native_tls::TlsConnectorBuilder::danger_accept_invalid_hostnames`],
+    /// [`rustls::ClientConfig::dangerous`]
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    pub fn danger_no_tls_verify(&mut self, no_tls_verify: bool) -> &mut Self {
+        self.no_tls_verify = no_tls_verify;
         self
     }
 
-    /// Return a new [`Client`] using a `native-tls` transport.
+    /// Makes a [`Client`] connection using the native TLS backend.
+    /// Forces TLS unless [`ClientBuilder::starttls()`] was called
+    ///
+    /// ```no_run
+    /// # use imap::ClientBuilder;
+    /// # {} #[cfg(feature = "rustls-tls")]
+    /// # fn main() -> Result<(), imap::Error> {
+    /// let client = ClientBuilder::new("imap.example.com", 993).native_tls()?;
+    /// # Ok(())
+    /// # }
+    /// ```
     #[cfg(feature = "native-tls")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "native-tls")))]
-    pub fn native_tls(&mut self) -> Result<Client<TlsStream<TcpStream>>> {
-        self.connect(|domain, tcp| {
-            let ssl_conn = TlsConnector::builder().build()?;
-            Ok(TlsConnector::connect(&ssl_conn, domain, tcp)?)
-        })
+    pub fn native_tls(&mut self) -> Result<Client<Connection>> {
+        if self.mode == ConnectionMode::Tcp {
```

Comment:
It still feels so weird to me that we force `.tls()` here, but _not_ for `.tls_kind()`.

Reply:
Not sure what you mean here.

Reply:
`Client::tls_kind` does not set `self.mode = Tls`. But `Client::native_tls` (which really just changes the kind) _does_ set `self.mode = Tls`. That feels inconsistent.

Reply:
I agree — the helpers also end up confusing because it's hard to tell that `tls` sets the _mode_, but `native_tls` sets the _kind_. Having just mode and kind is clearer I think.